### PR TITLE
Compile with ui library

### PR DESCRIPTION
**Pull request by m-masataka**
I want the method that capacitate a developer for compiling harbor with "Harbor UI Library".
I add the  ``COMPILEHARBORUIFLAG`` .  
If we execute ``make install`` with this flag, we can compile with LOCAL Harbor UI Library.

### CODE REVIEW

**Comment by steven-zou** [comment]
@m-masataka 

  Thanks for submitting PR to Harbor. From your comments, we still did not get what's your main purpose for creating this PR? Could you please explain more?

**Comment by m-masataka** [comment]
@steven-zou 
Thank you for replying me.
In Development Environment, I want to develop and build harbor with harbor-ui.
Unfortunately, current ``Makefile`` doesn't use LOCAL harbor ui library (``src/ui_ng/lib/ ``) but public harbor-ui's npm [package](https://www.npmjs.com/package/harbor-ui) when target ``compile_clarity:`` . (Is this correct?)
This specification is not developer friendly.
So I added the option what to use for make harbor with local harbor ui library(``src/ui_ng/lib/ ``).

**Comment by steven-zou** [comment]
@m-masataka 
  Sounds great! Actually your PR will help us a lot! Thanks!

@wy65701436 ,
  Please help to review the makefile code in this PR, thanks.

File: make/dev/npm-installer/entrypoint.sh [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+if [ ! -z "$npm_proxy" -a "$npm_proxy" != " " ]; then
+	npm config set proxy $npm_proxy
+fi
+
+cat ./package.json
+npm install
+./node_modules/.bin/ngc -p tsconfig-aot.json
+
+npm install
```
Comment (steven-zou) [inline make/dev/npm-installer/entrypoint.sh]: These lines seem duplicated ones, please check.

**Comment by wy65701436** [comment]
@m-masataka, it's definitely a good idea for the developer and helpful for us.

Is it possible to add an new target instead of modifying the existing one(compile_clarity)? This is because that our CI/CD is using Make to generate build, it may break our exising process.

File: Makefile [hunk 2]
```diff
@@ -279,8 +280,54 @@ compile_clarity:
...
+	@echo $(GOBASEPATH)
+	@echo $(GOBUILDPATH)
+	@$(DOCKERCMD) run --rm -v $(BUILDPATH):$(GOBUILDPATH) -w $(GOBUILDPATH_UI) $(GOBUILDIMAGE) $(GOIMAGEBUILD) -v -o $(GOBUILDMAKEPATH_UI)/$(UIBINARYNAME)
+	@echo "Done."
+
+	@echo "compiling binary for jobservice (golang image)..."
+	@$(DOCKERCMD) run --rm -v $(BUILDPATH):$(GOBUILDPATH) -w $(GOBUILDPATH_JOBSERVICE) $(GOBUILDIMAGE) $(GOIMAGEBUILD) -v -o $(GOBUILDMAKEPATH_JOBSERVICE)/$(JOBSERVICEBINARYNAME)
+	@echo "Done."
+
+compile_golangimage: compile_clarity
```
Comment (wy65701436) [inline Makefile]: what's the purpose to add one more compile_golangimage target?

Comment (m-masataka) [inline Makefile]: Sorry. I was mistaken.

File: make/dev/npm-installer/entrypoint.sh (hunk 1 above)
Comment (m-masataka) [inline make/dev/npm-installer/entrypoint.sh]: @steven-zou 
The first ``npm install & ngc -p tsconfig-aot.json`` means that set up the Harbor-UI library.
The second one means that install Harbor-UI library in ``src/ui_ng`` environment. 
If I didn't execute these commands, dependency errors occurred.

**Comment by steven-zou** [comment]
@m-masataka 

   Any updates for this PR?

**Comment by m-masataka** [comment]
@steven-zou 
No , that all.

**Comment by steven-zou** [comment]
@m-masataka 

  What I mean in my last comment is @wy65701436  has raised some comments to this PR and you might need to consider those comments.

**Comment by steven-zou** [comment]
@m-masataka 

  Thanks again for providing this excited functionality to Harbor community. As we're reaching the deadline of next new release, to avoid the possible influences on the releasing progress, we may hold this PR for a while and then merged back to master.

**Comment by m-masataka** [comment]
@steven-zou 
Thank you for your comments, and Sorry my late reply.
I've responded to @wy65701436 's comment and fixed Makefile.
I'll fetch and merge master to this branch. Is it OK?

**Comment by steven-zou** [comment]
@m-masataka 

  Sure, no problem as long as this PR keeps clear(no conflicts) when merging back to master.

File: Makefile [hunk 3]
```diff
@@ -286,8 +287,45 @@ compile_clarity:
...
 
+compile_clarity_harborui:
+	@echo "compiling binary for clarity ui with local harbor ui library..."
+	@if [ ! -e $(UINGPATH)/package.json.org ] ; then \
+		cp $(UINGPATH)/package.json $(UINGPATH)/package.json.org; \
+		sed -i -e "s/\"harbor-ui\": \"\^[0-9].\+\"/\"harbor-ui\": \"file:.\/lib\/\"/g" $(UINGPATH)/package.json; \
+	fi
+	@if [ "$(HTTPPROXY)" != "" ] ; then \
+		$(SEDCMD) -i 's/__proxy__/--proxy $(HTTPPROXY)/g' $(DOCKERFILE_BUILD_UI) ; \
+		$(DOCKERBUILD) -f $(DOCKERFILE_BUILD_UI) -t $(DOCKERIMAGENAME_CLATIRY):dev .; \
```
Comment (wy65701436) [inline Makefile]: please don't hard code the image tag here, you can use commit id like https://github.com/vmware/harbor/blob/master/make/photon/Makefile#L55 does.

File: Makefile [hunk 4]
```diff
@@ -286,8 +287,45 @@ compile_clarity:
...
+		$(DOCKERCMD) run --rm -v $(BUILDPATH)/src:$(CLARITYSEEDPATH) $(DOCKERIMAGENAME_CLATIRY):dev $(SHELL) $(CLARITYBUILDSCRIPT); \
+	fi
+	@if [ -e $(UINGPATH)/package.json.org ] ; then \
+		mv $(UINGPATH)/package.json.org $(UINGPATH)/package.json; \
+	fi
+	@echo "Done."
+
 compile_normal: compile_clarity compile_adminserver compile_ui compile_jobservice
 
+compile_harborui: compile_clarity_harborui
```
Comment (wy65701436) [inline Makefile]: The target name seems not tell the exactly jobs here, please give an reasonable one.

Can you give an example on how to use this target in Make?  You can refer to this doc. https://github.com/vmware/harbor/blob/master/docs/use_make.md

Comment (m-masataka) [reply Makefile]: thank you for review.
I think this name is fit at jobs here.
compile_harborui -> compile_clarity_with_local_harborui
Is it OK?

Comment (wy65701436) [reply Makefile]: It's not just compile UI, but adminserver/jobservice. 

compile_golangimage_with_local_library, is it OK? Again, please commit example and description in the use_make.md.

Comment (m-masataka) [reply Makefile]: OK! 
thanks!

**Pull request closed without merging**